src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs

```diff
@@ -29,6 +28,7 @@ public static IQueryable AsQueryable(this IEnumerable source)
             return EnumerableQuery.Create(enumType.GenericTypeArguments[0], source);
         }
 
+        [DynamicDependency("Where`1", typeof(Enumerable))]
```

Comment:
Can you point me to where this reflection happens? I assume we need to suppress linker warnings in that spot because DynamicDependency only keeps stuff but doesn't help linker to analyze anything (as opposed to DynamicallyAccessedMembers annotations that both keep stuff and help analysis).

I wonder whether we can just write the reflection in a way that linker can analyze instead.

From a cursory look at reflection use around here, there are some ridiculously bad patterns. (E.g. constructing a delegate to something and grabbing a method info of the delegate, instead of simply doing a `typeof(X).GetMethod()` that is both fast and easy to analyze, etc.).

Reply:
If this is about the pattern in EnumerableRewriter.VisitMethodCall, I'm not convinced that it can be made linker safe in general (and once we place a suppression on it, we'll get a warm and fuzzy feeling that it's safe).

This part is probably covered:

https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs#L39-L45

But this part looks pretty suspicious to me:

https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs#L48-L51

Are we convinced this part is also okay? Because if not, another solution is just to mark Queryable as not linker safe.

Reply:
@MichalStrehovsky I think in the reported issue the reflection call is at https://github.com/dotnet/runtime/blob/6072e4d3a7a2a1493f514cdf4be75a3d56580e84/src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs#L218 but it's possible there are more code paths using reflection in queryable implementation

Reply:
@MichalStrehovsky I discussed this with Eric and you're right the second part with `FindMethod` is not solved and honestly that one is basically impossible to solve fully - I can see us doing something which would solve it for this specific case, but the way the API is designed we probably can't ever get the right information into the right place (linker would basically have to understand expression tree nodes and be able to track some additional metadata about them).

My understanding is that this change is meant as a fix for an issue reported in 5.0 - regression compared to Blazor 3.2. It's not meant as a full solution to the Linq problem.

Reply:
> My understanding is that this change is meant as a fix for an issue reported in 5.0 - regression compared to Blazor 3.2. It's not meant as a full solution to the Linq problem.

Correct.

If you want to see the current linker warnings on this library, check out the suppressions file being added in @layomia's PR #40691:

https://github.com/dotnet/runtime/blob/751582618e20d512eb5d4ce09877bae24c094b58/src/libraries/System.Linq.Queryable/src/ILLink/ILLink.Suppressions.xml

There's 110 of them - so lots of work here to make the full library trim-safe.

> the second part with FindMethod is not solved and honestly that one is basically impossible to solve fully

The more I look at it, the more I think this kind of code could never truly be trim-safe. My understanding of that `FindMethod` case is that if you have 2 methods:

```C#

public static void Foo<T>(Expression<Func<T>> func);
public static void Foo<T>(Func<T> func);

```

And the current `Expression` is pointing at the first method, `FindMethod` will find the second method (since `Expression<>` gets [stripped during the rewrite](https://github.com/dotnet/runtime/blob/9c472545df220d7ae26329e1649f880c4ee127e8/src/libraries/System.Linq.Queryable/src/System/Linq/EnumerableRewriter.cs#L289-L302). How could we ever model this in the ILLinker?

And the problem gets worse as we expand to user code, because anyone could write an `ExpressionVisitor` which could do any sort of reflection policy.

Reply:
It's hard to make 20 year of legacy APIs trimming friendly.

I think we'll end up with a type-level trimming mode eventually (where entire types are kept and linker warnings are scaled back to what's unsafe when trimming whole types only) - it will not be as efficient at deleting stuff as member level trimming, but patterns like this would be safe in it.

It might well be that Queryable will only be safe in this more conservative trimming mode and will warn otherwise. Of course this code should be investigated - is the pattern match documented and are people relying on it, or is this for internal use somewhere etc.?